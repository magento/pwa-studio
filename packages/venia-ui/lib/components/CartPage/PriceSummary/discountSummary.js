import React, { Fragment } from 'react';
import { FormattedMessage } from 'react-intl';
import Price from '@magento/venia-ui/lib/components/Price';

import { useStyle } from '../../../classify';

const MINUS_SYMBOL = '-';

const DEFAULT_AMOUNT = {
    currency: 'USD',
    value: 0
};

/**
 * Reduces discounts array into a single amount.
 *
 * @param {Array} discounts
 */
const getDiscount = (discounts = []) => {
    // discounts from data can be null
    if (!discounts || !discounts.length) {
        return DEFAULT_AMOUNT;
    } else {
        return {
            currency: discounts[0].amount.currency,
            value: discounts.reduce(
                (acc, discount) => acc + discount.amount.value,
                0
            )
        };
    }
};

/**
 * A component that renders the discount summary line item.
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */
const DiscountSummary = props => {
    const classes = useStyle({}, props.classes);
    const discount = getDiscount(props.data);

    return discount.value ? (
        <Fragment>
            <span
                className={classes.lineItemLabel}
                data-cy="PriceSummary-DiscountSummary-label"
            >
                <FormattedMessage
                    id={'discountSummary.lineItemLabel'}
                    defaultMessage={'Discounts applied'}
                />
            </span>
            <span
                data-cy="DiscountSummary-discountValue"
                className={classes.price}
            >
                {MINUS_SYMBOL}
                <Price
                    value={discount.value}
                    currencyCode={discount.currency}
                />
            </span>
        </Fragment>
    ) : null;
};

export default DiscountSummary;
