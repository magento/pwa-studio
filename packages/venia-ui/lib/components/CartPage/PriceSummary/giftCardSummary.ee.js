import React, { Fragment } from 'react';
import { FormattedMessage } from 'react-intl';
import Price from '@magento/venia-ui/lib/components/Price';

import { useStyle } from '../../../classify';

const MINUS_SYMBOL = '-';

const DEFAULT_AMOUNT = {
    currency: 'USD',
    value: 0
};

/**
 * Reduces applied gift card amounts into a single amount.
 *
 * @param {Array} cards
 */
const getGiftCards = (cards = []) => {
    if (!cards.length) {
        return DEFAULT_AMOUNT;
    } else {
        return {
            currency: cards[0].applied_balance.currency,
            value: cards.reduce(
                (acc, card) => acc + card.applied_balance.value,
                0
            )
        };
    }
};

/**
 * A component that renders the gift card summary line item.
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */
export default props => {
    const classes = useStyle({}, props.classes);
    const cards = getGiftCards(props.data);

    return cards.value ? (
        <Fragment>
            <span className={classes.lineItemLabel}>
                <FormattedMessage
                    id={'giftCardSummary.lineItemLabel'}
                    defaultMessage={'Gift Card(s) applied'}
                />
            </span>
            <span className={classes.price}>
                {MINUS_SYMBOL}
                <Price value={cards.value} currencyCode={cards.currency} />
            </span>
        </Fragment>
    ) : null;
};
