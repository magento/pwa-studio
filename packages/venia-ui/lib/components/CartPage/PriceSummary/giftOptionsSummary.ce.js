/**
 * Gift Options Prices are an AC-only feature.
 * Here in MOS, don't render the component at all.
 */
export default () => null;
