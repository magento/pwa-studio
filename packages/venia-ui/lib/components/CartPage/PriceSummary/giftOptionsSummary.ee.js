import React, { Fragment } from 'react';
import { FormattedMessage } from 'react-intl';

import { useStyle } from '@magento/venia-ui/lib/classify';
import Price from '@magento/venia-ui/lib/components/Price';

/**
 * A component that renders the gift options summary line item.
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */
export default props => {
    const classes = useStyle({}, props.classes);
    const giftOptions = props?.data?.printed_card || {};

    if (!giftOptions || !giftOptions.value) {
        return null;
    }

    return (
        <Fragment>
            <span
                className={classes.lineItemLabel}
                data-cy="PriceSummary-GiftOptionsSummary-label"
            >
                <FormattedMessage
                    id={'giftOptionsSummary.lineItemLabel'}
                    defaultMessage={'Printed Card'}
                />
            </span>
            <span className={classes.price}>
                <Price
                    value={giftOptions.value}
                    currencyCode={giftOptions.currency}
                />
            </span>
        </Fragment>
    );
};
