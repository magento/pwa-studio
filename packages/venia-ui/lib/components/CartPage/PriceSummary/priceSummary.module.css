.root {
}

.errorText {
    color: rgb(var(--venia-global-color-error));
    line-height: var(--venia-global-lineHeight-300);
}

.lineItems {
    display: grid;
    grid-gap: 0.75rem;
    grid-template-columns: 1fr auto;
    line-height: 1.5rem;
}

.lineItemLabel {
    justify-self: start;
}

.price {
    justify-self: end;
}

.totalLabel {
    composes: lineItemLabel;
    font-weight: var(--venia-global-fontWeight-semibold);
}

.totalPrice {
    composes: price;
    font-weight: var(--venia-global-fontWeight-semibold);
}

.priceUpdating {
    composes: price;
    opacity: 0;
}

.checkoutButton_container {
    align-items: center;
    display: inline-flex;
    justify-content: center;
    margin-top: 1rem;
    min-height: 5.5rem;
    width: 100%;
}
