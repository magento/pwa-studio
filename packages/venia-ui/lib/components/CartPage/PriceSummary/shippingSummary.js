import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import Price from '@magento/venia-ui/lib/components/Price';

import { useStyle } from '../../../classify';
/**
 * A component that renders the shipping summary line item after address and
 * method are selected
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */
const ShippingSummary = props => {
    const classes = useStyle({}, props.classes);
    const { data, isCheckout } = props;
    const { formatMessage } = useIntl();

    // Don't render estimated shipping until an address has been provided and
    // a method has been selected.
    if (!data.length || !data[0].selected_shipping_method) {
        return null;
    }

    const shipping = data[0].selected_shipping_method.amount;

    const shippingLabel = isCheckout
        ? formatMessage({
              id: 'shippingSummary.shipping',
              defaultMessage: 'Shipping'
          })
        : formatMessage({
              id: 'shippingSummary.estimatedShipping',
              defaultMessage: 'Estimated Shipping'
          });

    // For a value of "0", display "FREE".
    const price = shipping.value ? (
        <Price value={shipping.value} currencyCode={shipping.currency} />
    ) : (
        <span>
            <FormattedMessage id={'global.free'} defaultMessage={'FREE'} />
        </span>
    );

    return (
        <>
            <span className={classes.lineItemLabel}>{shippingLabel}</span>
            <span className={classes.price}>{price}</span>
        </>
    );
};

export default ShippingSummary;
