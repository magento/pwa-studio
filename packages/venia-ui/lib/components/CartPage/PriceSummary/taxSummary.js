import React from 'react';
import { useIntl } from 'react-intl';
import Price from '@magento/venia-ui/lib/components/Price';

import { useStyle } from '../../../classify';
/**
 * Reduces applied tax amounts into a single amount.
 *
 * @param {Array} applied_taxes
 */
const getEstimatedTax = (applied_taxes = []) => {
    return {
        currency: applied_taxes[0].amount.currency,
        value: applied_taxes.reduce((acc, tax) => acc + tax.amount.value, 0)
    };
};

/**
 * A component that renders the tax summary line item.
 *
 * @param {Object} props.classes
 * @param {Object} props.data query response data
 */
const TaxSummary = props => {
    const classes = useStyle({}, props.classes);
    const { data, isCheckout } = props;
    const { formatMessage } = useIntl();

    // Don't render estimated taxes until an address has been provided which
    // causes the server to apply a tax value to the cart.
    if (!data.length) {
        return null;
    }

    const taxLabel = isCheckout
        ? formatMessage({
              id: 'taxSummary.tax',
              defaultMessage: 'Tax'
          })
        : formatMessage({
              id: 'taxSummary.estimatedTax',
              defaultMessage: 'Estimated Tax'
          });

    const tax = getEstimatedTax(props.data);

    return (
        <>
            <span className={classes.lineItemLabel}>{taxLabel}</span>
            <span className={classes.price}>
                <Price value={tax.value} currencyCode={tax.currency} />
            </span>
        </>
    );
};

export default TaxSummary;
