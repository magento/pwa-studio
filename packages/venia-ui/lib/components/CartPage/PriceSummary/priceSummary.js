import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import Price from '@magento/venia-ui/lib/components/Price';
import { usePriceSummary } from '@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary';
import Button from '../../Button';
import { useStyle } from '../../../classify';
import defaultClasses from './priceSummary.module.css';
import DiscountSummary from './discountSummary';
import GiftCardSummary from './giftCardSummary';
import ShippingSummary from './shippingSummary';
import TaxSummary from './taxSummary';

/**
 * A child component of the CartPage component.
 * This component fetches and renders cart data, such as subtotal, discounts applied,
 * gift cards applied, tax, shipping, and cart total.
 *
 * @param {Object} props
 * @param {Object} props.classes CSS className overrides.
 * See [priceSummary.module.css]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/PriceSummary/priceSummary.module.css}
 * for a list of classes you can override.
 *
 * @returns {React.Element}
 *
 * @example <caption>Importing into your project</caption>
 * import PriceSummary from "@magento/venia-ui/lib/components/CartPage/PriceSummary";
 */
const PriceSummary = props => {
    const { isUpdating } = props;
    const classes = useStyle(defaultClasses, props.classes);
    const talonProps = usePriceSummary();

    const {
        handleProceedToCheckout,
        hasError,
        hasItems,
        isCheckout,
        isLoading,
        flatData
    } = talonProps;
    const { formatMessage } = useIntl();

    if (hasError) {
        return (
            <div className={classes.root}>
                <span className={classes.errorText}>
                    <FormattedMessage
                        id={'priceSummary.errorText'}
                        defaultMessage={
                            'Something went wrong. Please refresh and try again.'
                        }
                    />
                </span>
            </div>
        );
    } else if (!hasItems) {
        return null;
    }

    const { subtotal, total, discounts, giftCards, taxes, shipping } = flatData;

    const isPriceUpdating = isUpdating || isLoading;
    const priceClass = isPriceUpdating ? classes.priceUpdating : classes.price;
    const totalPriceClass = isPriceUpdating
        ? classes.priceUpdating
        : classes.totalPrice;

    const totalPriceLabel = isCheckout
        ? formatMessage({
              id: 'priceSummary.total',
              defaultMessage: 'Total'
          })
        : formatMessage({
              id: 'priceSummary.estimatedTotal',
              defaultMessage: 'Estimated Total'
          });

    const proceedToCheckoutButton = !isCheckout ? (
        <div className={classes.checkoutButton_container}>
            <Button
                disabled={isPriceUpdating}
                priority={'high'}
                onClick={handleProceedToCheckout}
            >
                <FormattedMessage
                    id={'priceSummary.checkoutButton'}
                    defaultMessage={'Proceed to Checkout'}
                />
            </Button>
        </div>
    ) : null;

    return (
        <div className={classes.root} data-cy="PriceSummary-root">
            <div className={classes.lineItems}>
                <span className={classes.lineItemLabel}>
                    <FormattedMessage
                        id={'priceSummary.lineItemLabel'}
                        defaultMessage={'Subtotal'}
                    />
                </span>
                <span
                    data-cy="PriceSummary-subtotalValue"
                    className={priceClass}
                >
                    <Price
                        value={subtotal.value}
                        currencyCode={subtotal.currency}
                    />
                </span>
                <DiscountSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={discounts}
                />
                <GiftCardSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={giftCards}
                />
                <TaxSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={taxes}
                    isCheckout={isCheckout}
                />
                <ShippingSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={shipping}
                    isCheckout={isCheckout}
                />
                <span className={classes.totalLabel}>{totalPriceLabel}</span>
                <span
                    data-cy="PriceSummary-totalValue"
                    className={totalPriceClass}
                >
                    <Price value={total.value} currencyCode={total.currency} />
                </span>
            </div>
            {proceedToCheckoutButton}
        </div>
    );
};

export default PriceSummary;
