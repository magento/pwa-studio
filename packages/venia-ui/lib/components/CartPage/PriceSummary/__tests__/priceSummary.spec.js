import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { usePriceSummary } from '@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary';
import PriceSummary from '../priceSummary';

jest.mock('../../../../classify');
jest.mock('react-router-dom', () => ({
    Link: jest.fn(() => 'Proceed to Checkout')
}));
jest.mock('@apollo/client', () => {
    const runQuery = jest.fn();
    const queryResult = {
        data: {
            cart: {
                items: [
                    {
                        quantity: 1
                    }
                ],
                applied_gift_cards: [],
                shipping_addresses: [
                    {
                        selected_shipping_method: {
                            amount: {
                                value: 0,
                                currency: 'USD'
                            }
                        }
                    }
                ],
                prices: {
                    subtotal_excluding_tax: {
                        currency: 'USD',
                        value: 11
                    },
                    grand_total: {
                        currency: 'USD',
                        value: 10
                    },
                    discounts: [
                        {
                            amount: {
                                value: 1,
                                currency: 'USD'
                            }
                        }
                    ],
                    applied_taxes: [
                        {
                            amount: {
                                value: 0,
                                currency: 'USD'
                            }
                        }
                    ]
                }
            }
        },
        error: null,
        loading: false
    };
    const useLazyQuery = jest.fn(() => [runQuery, queryResult]);

    return {
        gql: jest.fn(),
        useLazyQuery
    };
});

const defaultTalonProps = {
    hasError: false,
    hasItems: true,
    isCheckout: false,
    isLoading: false,
    flatData: {
        subtotal: { currency: 'USD', value: 3.5 },
        total: { currency: 'USD', value: 8.5 },
        discounts: null,
        giftCards: [],
        taxes: [],
        shipping: { currency: 'USD', value: 5 }
    }
};

jest.mock(
    '@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary',
    () => ({
        usePriceSummary: jest.fn().mockReturnValue({
            hasError: false,
            hasItems: true,
            isCheckout: false,
            isLoading: false,
            flatData: {
                subtotal: { currency: 'USD', value: 3.5 },
                total: { currency: 'USD', value: 8.5 },
                discounts: null,
                giftCards: [],
                taxes: [],
                shipping: { currency: 'USD', value: 5 }
            }
        })
    })
);

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

test('renders PriceSummary correctly on cart page', () => {
    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders PriceSummary correctly on checkout page', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        isCheckout: true
    });

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders an error state if query fails', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        hasError: true
    });

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders summary with loading state if query is loading', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        isLoading: true
    });

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders nothing if query returns no items', () => {
    usePriceSummary.mockReturnValueOnce({
        ...defaultTalonProps,
        hasItems: false
    });

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});
