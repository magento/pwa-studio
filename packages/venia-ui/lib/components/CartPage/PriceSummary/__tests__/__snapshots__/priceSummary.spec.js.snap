// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PriceSummary correctly on cart page 1`] = `
<div
  className="root"
>
  <div
    className="lineItems"
  >
    <span
      className="lineItemLabel"
    >
      <mock-FormattedMessage
        defaultMessage="Subtotal"
        id="priceSummary.lineItemLabel"
      />
    </span>
    <span
      className="price"
    >
      <span>
        $
      </span>
      <span>
        3
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
    <span
      className="totalLabel"
    >
      Estimated Total
    </span>
    <span
      className="totalPrice"
    >
      <span>
        $
      </span>
      <span>
        8
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
  </div>
  <div
    className="checkoutButton_container"
  >
    <button
      className="root_highPriority"
      disabled={false}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        <mock-FormattedMessage
          defaultMessage="Proceed to Checkout"
          id="priceSummary.checkoutButton"
        />
      </span>
    </button>
  </div>
</div>
`;

exports[`renders PriceSummary correctly on checkout page 1`] = `
<div
  className="root"
>
  <div
    className="lineItems"
  >
    <span
      className="lineItemLabel"
    >
      <mock-FormattedMessage
        defaultMessage="Subtotal"
        id="priceSummary.lineItemLabel"
      />
    </span>
    <span
      className="price"
    >
      <span>
        $
      </span>
      <span>
        3
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
    <span
      className="totalLabel"
    >
      Total
    </span>
    <span
      className="totalPrice"
    >
      <span>
        $
      </span>
      <span>
        8
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
  </div>
</div>
`;

exports[`renders an error state if query fails 1`] = `
<div
  className="root"
>
  <span
    className="errorText"
  >
    <mock-FormattedMessage
      defaultMessage="Something went wrong. Please refresh and try again."
      id="priceSummary.errorText"
    />
  </span>
</div>
`;

exports[`renders nothing if query returns no items 1`] = `null`;

exports[`renders summary with loading state if query is loading 1`] = `
<div
  className="root"
>
  <div
    className="lineItems"
  >
    <span
      className="lineItemLabel"
    >
      <mock-FormattedMessage
        defaultMessage="Subtotal"
        id="priceSummary.lineItemLabel"
      />
    </span>
    <span
      className="priceUpdating"
    >
      <span>
        $
      </span>
      <span>
        3
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
    <span
      className="totalLabel"
    >
      Estimated Total
    </span>
    <span
      className="priceUpdating"
    >
      <span>
        $
      </span>
      <span>
        8
      </span>
      <span>
        .
      </span>
      <span>
        50
      </span>
    </span>
  </div>
  <div
    className="checkoutButton_container"
  >
    <button
      className="root_highPriority"
      disabled={true}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        <mock-FormattedMessage
          defaultMessage="Proceed to Checkout"
          id="priceSummary.checkoutButton"
        />
      </span>
    </button>
  </div>
</div>
`;
